File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    def combinationSum2(self, cand: List[int], tar: int) -> List[List[int]]:
        j = 0
        res = []
        temp = []
        cand.sort()
        def fun(j, tar):
            if tar == 0:
                res.append(temp.copy())
                return
            elif tar < 0:
                return
            prev = -1
            for i in range(j, len(cand)):
                if prev == cand[i]:
                    continue
                temp.append(cand[i])
                fun(i+1, tar - cand[i])
                temp.pop()
                prev = cand[i]
        fun(0, tar)
        return res
                

File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& cand, int tar) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(begin(cand), end(cand));
        fun(cand, tar, ans, temp ,0);
        return ans;
    }
private:
    void fun(vector<int>& cand, int k,vector<vector<int>> &ans, vector<int> &temp, int j) {
        if (k == 0) {
            ans.push_back(temp);
            return;
        }
        if (k < 0) return;
        int prev = -1;
        for (int i = j;i < cand.size();i++) {
            if (cand[i] == prev) continue;
            temp.push_back(cand[i]);
            fun(cand, k - cand[i], ans, temp, i+1);
            temp.pop_back();
            prev = cand[i];
        }
       
    }
};